### DISCUSSION TITLE
Quoted values in csv and line feeds/carriage returns in the middle of values ?

### DISCUSSION THREAD

**Discussion by SqlBenjamin** [JoshClose/CsvHelper#2013] [General]
So, I wrote my own code to parse csv I'm getting in memory (does not exist on disk anywhere) and then heard about this solution...however, I know I've had to handle line feeds and/or carriage returns in the middle of values (sometimes multiples) along with a couple other things and I'm not sure if CsvHelper handles that. For example, let's say I have the following:

"Column1","Column2","Column3"
"blah","hello","some other value"
"He said "what's up?"","this is a
multi line
value with a " in it","Josh,
this last column also has a line feed in it"
"value1","value2","value3"

Does CsvHelper properly get the 3 rows + 1 header row? It's comma delimited but values are between double quotes...but there can be quotes inside of the values as well as values with line feeds. If CsvHelper is easily capable of parsing this, is there some example I can see? (Where can I find some documentation/examples?)

I do see that there is also a CsvDataReader to load a datatable which is intereting because currently I use the header row in conjunction with a table definition - where the names match I use the type and nullable properties listed otherwise just use string and nullable values - to create a datatable, then populate that table with the rows. I doubt that class allows for that level of 'dynamic' but it would be pretty awesome if it does.

Thanks in advance for the answers/pointers.

**Comment by JoshClose** [comment]
That's not a valid CSV file. Check out the format from RFC 4180. https://www.rfc-editor.org/rfc/rfc4180

If you have an invalid format like that, there are some things you can try to still read it, but there isn't a way to know that a double quote in the middle of a field isn't then end of the field.

**Reply by SqlBenjamin** [reply]
Yeah, I figured it wasn't valid...but I'm still getting that garbage from the service I'm hitting. I'll create a bug for the service and if they ever fix it. If they do eventually fix it and only produce valid csv formats, I'll see if CsvHelper can be used to do what I need:
I use the header row to build the columns of a data table and then use the lines to create the rows (based on their ordinal position). Does CsvHelper have a quick and easy way to build a data table like that?

**Reply by JoshClose** [reply]
You could give this a try. https://joshclose.github.io/CsvHelper/examples/csvdatareader/
